### PR TITLE
feat: logs parsing pipeline support in opamp

### PR DESCRIPTION
**Pull request by nityanandagohain**
PR for https://github.com/SigNoz/signoz/issues/1779

This PR only adds the logic on how log pipelines will be sent to opamp and how it will be handled by agent conf.

### CODE REVIEW

File: pkg/query-service/app/opamp/logspipeline_test.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			err := BuildLogParsingProcessors(test.agentConf, test.pipelineProcessor)
+			So(err, ShouldBeNil)
+			So(test.agentConf, ShouldResemble, test.outputConf)
+		})
+	}
+
+}
+
+var BuildLogsPipelineTestData = []struct {
+	Name             string
```
Comment (nityanandagohain) [inline pkg/query-service/app/opamp/logspipeline_test.go:103]: I would love to have some eyes on these test cases.

**Review by srikanthccv — COMMENTED** [review]
reviewing the logic part of the `buildLogsProcessors ` at the moment because I am still finding it hard to follow the rest with the loose `interfaces{}` use for components.

File: pkg/query-service/app/opamp/logspipeline_test.go (hunk 1 above)
Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline_test.go:103]: Yes so  the array of interfaces is nothing but an array of strings.

So my expected behaviour is that 
* if there is no existing `log parsing processor` in the processor array i.e `currentPipeline` add it to the log parsing processors to the beginning.
 eg:- currentPipeline is ` filter, batch`  and I want to add `logparser_a , logparser_b` then the final result is  `logparser_a , logparser_b, filter, batch`
 
 * If someone modifies the currentPipeline to `filter, logparser_a , logparser_b, batch` and the new version of log parsing pipeline contains only `logparser_a` then final result is `filter, logparser_a , batch`

* If someone adds a processor in between and currentPipeline is `filter, logparser_a , processor2, logparser_b, batch` and log parsing pipelines contains ` logparser_a , logparser_b, logparser_c` then `logparser_c` will be added by taking `logparser_b` as anchor the final result is `filter, logparser_a , processor2, logparser_b, logparser_c, batch`

**Comment by nityanandagohain** [comment]
Have added a reply @srikanthccv  https://github.com/SigNoz/signoz/pull/2456#discussion_r1138093466

File: pkg/query-service/app/opamp/logspipeline_test.go (hunk 1 above)
Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline_test.go:103]: Okay, I am checking your examples, and I have updated the interfaces to structs.

Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline_test.go:103]:
```
{
Name:             "rearrange pipelines",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch"},
logsPipeline:     []string{constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a"},
expectedPipeline: []string{"processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", "batch"},
},
```

Here a should are present next to b taking b as an anchor

Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline_test.go:103]:
```
{
Name:             "rearrange pipelines with new processor",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch"},
logsPipeline:     []string{constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_c"},
expectedPipeline: []string{"processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_c", "batch"},
},
```

Here also we should take `_b` as the anchor and add the remaining ones.

Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline_test.go:103]:
```
{
Name:             "last to first",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", "processor4", constants.LogsPPLPfx + "_b", "batch", constants.LogsPPLPfx + "_c"},
logsPipeline:     []string{constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_b"},
expectedPipeline: []string{"processor1", "processor2", "processor3", "processor4", "batch", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_b"},
},
```
Take `_c` as anchor

Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline_test.go:103]:
```
{
Name:             "multiple rearrange pipelines",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch", constants.LogsPPLPfx + "_c", "processor4", "processor5", constants.LogsPPLPfx + "_d", "processor6", "processor7"},
logsPipeline:     []string{constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e"},
expectedPipeline: []string{"processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", "batch", "processor4", "processor5", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e", "processor6", "processor7"},
},
```

here also anchor should be taken on what is found if something is left then add it after it.

Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline_test.go:103]:
```
{
Name:             "multiple rearrange with new pipelines",
currentPipeline:  []string{"processor1", "processor2", constants.LogsPPLPfx + "_a", "processor3", constants.LogsPPLPfx + "_b", "batch", constants.LogsPPLPfx + "_c", "processor4", "processor5", constants.LogsPPLPfx + "_d", "processor6", "processor7"},
logsPipeline:     []string{constants.LogsPPLPfx + "_z", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e"},
expectedPipeline: []string{constants.LogsPPLPfx + "_z", "processor1", "processor2", "processor3", constants.LogsPPLPfx + "_b", constants.LogsPPLPfx + "_a", "batch", "processor4", "processor5", constants.LogsPPLPfx + "_d", constants.LogsPPLPfx + "_c", constants.LogsPPLPfx + "_e", "processor6", "processor7"},
},
```

Here also we go in order of `logsPipeline` and then if something is not in order we unmark it and add it later

**Comment by nityanandagohain** [comment]
Have updated the logic and added the new test cases for processor ordering, please have a look again.

**Comment by nityanandagohain** [comment]
@srikanthccv  please have a look at this when you get some time.

File: pkg/query-service/app/opamp/logspipeline.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	if len(agents) == 0 {
+		return confHash, fmt.Errorf("no agents available at the moment")
+	}
+
+	for _, agent := range agents {
+		config := agent.EffectiveConfig
+		c, err := yaml.Parser().Unmarshal([]byte(config))
+		if err != nil {
+			return confHash, err
+		}
```
Comment (srikanthccv) [inline pkg/query-service/app/opamp/logspipeline.go:36]: I am hoping there will be a PR for this as part of https://github.com/SigNoz/engineering-pod/issues/963 because I see this as a poorly thought out where the iterating over agents and updating it by `SendToAgent` multiple places. What was the manager designed to do?

File: pkg/query-service/app/opamp/logspipeline.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		return nil, err
+	}
+	p := otelPipeline{}
+	if err := json.Unmarshal(b, &p); err != nil {
+		return nil, err
+	}
+	return &p, nil
+}
+
+func buildLogsProcessors(current []string, logsParserPipeline []string) ([]string, error) {
```
Comment (srikanthccv) [inline pkg/query-service/app/opamp/logspipeline.go:140]: I know you originally used the same version of sampling rules logic, but does it have to be complicated as this for your case since we are going with the fact all SigNoz added processors would be a single group, and in the order, they are defined in the UI.

Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline.go:140]: So even if we consider that everything will be together and processors can be added at the beginning or towards the end of the group, this logic handles them. 

I agree that if we spend some more time we might be able to reduce the complexity where we don't need to consider anything in between but overall the logic will be something similar.

File: pkg/query-service/app/opamp/logspipeline.go (hunk 2 above)
Comment (nityanandagohain) [reply pkg/query-service/app/opamp/logspipeline.go:36]: Yeah, I had a discussion with Amol, we can merge this into one. But It can be added as an enhancement to this PR as different things are happening while iterating over config for different use cases.

**Review by srikanthccv — APPROVED** [review]

**Comment by srikanthccv** [comment]
Please squash next time.

**Comment by nityanandagohain** [comment]
Sure @srikanthccv

**Comment by nityanandagohain** [comment]
Yeah correct, will take care of this from the next PR's.
